index.d.ts

```diff
@@ -0,0 +1,193 @@
+import * as http from 'http';
+
+export = Router;
+
+declare namespace Router {
+
+  interface RouterOptions {
+    strict?: boolean;
+    caseSensitive?: boolean;
+    mergeParams?: boolean;
+  }
+
+  interface Layer {
+    name: string,
+    handle: RequestHandler | ErrorRequestHandler,
+    handle_request: RequestHandler,
+    handle_error: ErrorRequestHandler,
+    match: (path: string) => boolean,
+  }
+
+  interface IncomingRequest extends http.IncomingMessage {
+    originalUrl?: string,
+    params?: {
+      [key: string]: any,
+    },
+  }
+
+  interface RoutedRequest extends IncomingRequest {
+    baseUrl: string,
+    next?: NextFunction,
+    route?: IRoute
+  }
+
+  type RequestParamHandler = (req: IncomingRequest, res: http.ServerResponse, next: NextFunction, value: string, name: string) => void;
+
+  type RouteHandler = (req: RoutedRequest, res: http.ServerResponse, next: NextFunction) => void;
+  type RequestHandler = (req: IncomingRequest, res: http.ServerResponse, next: NextFunction) => void;
+
+  type NextFunction = (err?: Error | "route" | "router") => void;
+  type Callback = (err?: Error) => void;
+
+  type ErrorRequestHandler = (err: Error, req: IncomingRequest, res: http.ServerResponse, next: NextFunction) => void;
```

Comment:
We may not want the `err` to be typed as just a `Error` object, as in the 1.x of this router, there is no enforcement on that; using a middleware you didn't write could have a string or plain object here, which I've seen a lot. We wouldn't want someone's TS code to think they don't need some kind of guard on the `err` value.

Reply:
What's more appropriate type for that then? `any`?

Reply:
That is probably fine. It does have to be truthy, though, so not sure if that helps at all.

Reply:
Changed it to `any`.